source/lib/cdk-textract-stack.ts

```diff
@@ -122,6 +127,276 @@ export class CdkTextractStack extends cdk.Stack {
+        })
+      );
+
+      kendraRole.addToPolicy(
+        new iam.PolicyStatement({
+          effect: iam.Effect.ALLOW,
+          actions: ["s3:GetObject", "s3:ListBucket"],
+          resources: [
+            covidDataBucket.bucketArn,
+            `${covidDataBucket.bucketArn}/*`
+          ]
+        })
+      );
+
+      kendraRoleArn = kendraRole.roleArn;
+
+      const onEventCreateKendraIndexLambda = new lambda.Function(this, this.resourceName('OnEventKendraIndexCreator'), {
+        code: lambda.Code.fromAsset('lambda/kendraIndexCreator'),
+        description: 'onEvent handler for creating Kendra index',
+        runtime: lambda.Runtime.PYTHON_3_7,
+        handler: 'lambda_function.lambda_handler',
+        timeout: Duration.minutes(15),
+        environment: {
+          KENDRA_ROLE_ARN: kendraRole.roleArn,
+          KMS_KEY_ID: kendraKMSKey.keyId,
+          KENDRA_INDEX_CLIENT_TOKEN: this.uuid.toLowerCase()
+        }
+      });
+
+      kendraKMSKey.grantEncryptDecrypt(onEventCreateKendraIndexLambda);
+      onEventCreateKendraIndexLambda.addLayers(boto3Layer);
+
+      onEventCreateKendraIndexLambda.addToRolePolicy(
+        new iam.PolicyStatement({
+          effect: iam.Effect.ALLOW,
+          actions: ["kendra:CreateIndex","kendra:DeleteIndex","kendra:DescribeIndex","kendra:TagResource","kms:ListKeys",
+          "kms:ListAliases","kms:DescribeKey","kms:CreateGrant"],
+          resources: ["*"]
+        })
+      );
+
+      onEventCreateKendraIndexLambda.addToRolePolicy(
+        new iam.PolicyStatement({
+          effect: iam.Effect.ALLOW,
+          actions: ["iam:PassRole"],
+          resources: [kendraRole.roleArn]
+        })
+      );
+
+      const isCompleteKendraIndexLambda = new lambda.Function(this, this.resourceName('isCompleteKendraIndexPoller'),{
+        code: lambda.Code.fromAsset('lambda/kendraIndexPoller'),
+        description: 'isComplete handler to check for Kendra Index creation',
+        runtime: lambda.Runtime.PYTHON_3_7,
+        handler: 'lambda_function.lambda_handler',
+        timeout: Duration.minutes(15),
+      });
+
+      isCompleteKendraIndexLambda.addLayers(boto3Layer);
+      //kendraKMSKey.grantEncryptDecrypt(isCompleteKendraIndexLambda);
+
+      isCompleteKendraIndexLambda.addToRolePolicy(
+        new iam.PolicyStatement({
+          effect: iam.Effect.ALLOW,
+          actions: ["kendra:DescribeIndex"],
+          resources: ["*"]
+        })
+      );
+
+      isCompleteKendraIndexLambda.addToRolePolicy(
+        new iam.PolicyStatement({
+          effect: iam.Effect.ALLOW,
+          actions: ["iam:PassRole"],
+          resources: [kendraRole.roleArn]
+        })
+      );
+
+      const kendraIndexProvider = new cr.Provider(this, this.resourceName('KendraIndexProvider'), {
+        onEventHandler: onEventCreateKendraIndexLambda,
+        isCompleteHandler: isCompleteKendraIndexLambda,
+        totalTimeout: Duration.hours(1),
```

Comment:
Suppose Kendra takes about a little over 1 hour to get creeated. The timneout may occur at 1 hour but how do we amke sure no half-created solution (with Kendra index ) is not running?

Reply:
Should I increase the timeout?

Reply:
* If totalTimeout expires and the operation has not yet completed, the framework will submit a "FAILED" response with the message "Operation timed out". https://docs.aws.amazon.com/cdk/api/latest/docs/custom-resources-readme.html#handling-lifecycle-events-onevent
* Before the index creation is complete, the isCompleteHandler does not exit and the stack remains in status - CREATE_IN_PROGRESS.